include/libpmemobj++/experimental/radix.hpp

```diff
@@ -0,0 +1,1813 @@
+{
+	auto key = bytes_view(k);
+
+	auto n = root;
+	auto child_slot = &root;
+	while (n && !n.is_leaf()) {
+		if (n->byte == key.size() && n->bit == 8 - SLICE)
+			child_slot = &n->leaf;
+		else if (n->byte > key.size())
+			return end();
+		else
+			child_slot = &n->child[slice_index(key.data()[n->byte],
+							   n->bit)];
+
+		n = *child_slot;
+	}
+
+	if (!n)
+		return end();
+
+	if (!keys_equal(key, bytes_view(n.get_leaf()->key())))
+		return end();
+
+	return const_iterator(child_slot);
+}
+
+template <typename Key, typename Value>
+void
+radix_tree<Key, Value>::clear()
+{
+	if (size() != 0)
+		erase(begin(), end());
+}
+
+template <typename Key, typename Value>
+typename radix_tree<Key, Value>::iterator
+radix_tree<Key, Value>::erase(iterator pos)
+{
+	auto pop = pool_base(pmemobj_pool_by_ptr(this));
+
+	transaction::run(pop, [&] {
+		auto *leaf = pos.node;
+		auto parent = leaf->get_leaf()->parent;
+
+		delete_persistent<radix_tree::leaf>(leaf->get_leaf());
+
+		size_--;
+
+		/* was root */
+		if (!parent) {
+			root = nullptr;
+			pos = begin();
+			return;
+		}
+
+		++pos;
+		*leaf = nullptr;
+
+		/* Compress the tree vertically. */
+		auto n = parent;
+		parent = n->parent;
+		tagged_node_ptr only_child = nullptr;
+		for (size_t i = 0; i < SLNODES; i++) {
+			if (n->child[i]) {
+				if (only_child) {
+					/* more than one child */
+					return;
+				}
+				only_child = n->child[i];
+			}
+		}
+
+		if (only_child && n->leaf) {
+			/* There are actually 2 "childred" so we can't compress.
+			 */
+			return;
+		} else if (n->leaf)
+			only_child = n->leaf;
+
+		assert(only_child);
+		parent_ref(only_child) = n->parent;
+
+		auto *child_slot = parent
+			? const_cast<tagged_node_ptr *>(
+				  parent->find_child(n).operator->())
+			: &root;
+		*child_slot = only_child;
+
+		/* If iterator now points to only_child it mus be updated
+		 * (otherwise it would have reference to freed node). */
+		if (pos.node && *pos.node == only_child)
+			pos.node = child_slot;
+
+		delete_persistent<radix_tree::node>(n.get_node());
+	});
+
+	return pos;
+}
+
+template <typename Key, typename Value>
+void
+radix_tree<Key, Value>::erase(iterator first, iterator last)
+{
+	auto pop = pool_base(pmemobj_pool_by_ptr(this));
+
+	transaction::run(pop, [&] {
+		while (first != last)
+			first = erase(first);
+	});
+}
+
+template <typename Key, typename Value>
+typename radix_tree<Key, Value>::size_type
+radix_tree<Key, Value>::erase(const_key_reference k)
+{
+	auto it = find(k);
+
+	if (it == end())
+		return 0;
+
+	erase(it);
+
+	return 1;
+}
+
+template <typename Key, typename Value>
+typename radix_tree<Key, Value>::const_iterator
+radix_tree<Key, Value>::lower_bound(const_key_reference k) const
+{
+	auto key = bytes_view(k);
+
+	auto n = root;
+	decltype(n) prev;
+	decltype(n) *child_slot = nullptr;
+
+	if (!root)
+		return end();
+
+	while (n && !n.is_leaf()) {
+		prev = n;
+
+		if (n->byte == key.size() && n->bit == 8 - SLICE)
```

Comment:
change all of the `n->nbit == 8 - SLICE` checks if something descriptive. (node_has_embedded_leaf or something?)